test/test.pl

```diff
@@ -350,6 +352,41 @@ sub test_view
     testv $opts, "./compare_sam.pl range.tmp range.out";
 }
 
+# Tests CRAM's ability to correctly preserve MD and NM, irrespective of whether
+# they are correct.
+sub test_MD
+{
+    my ($opts) = @_;
+
+    foreach my $sam (glob("*#MD*.sam")) {
+        my ($base, $ref) = ($sam =~ /((.*)#.*)\.sam/);
+        $ref .= ".fa";
+
+        my $bam  = "$base.tmp.bam";
+        my $cram = "$base.tmp.cram";
+
+        print "\ntest_MD testing $sam, ref $ref:\n";
+        $test_view_failures = 0;
+        $cram = "$base.tmp.cram";
+
+	# Forcibly store MD and NM and don't auto-generate.
+	# ALL NM/MD should match and be present only when originally present
+        testv $opts, "./test_view -o store_nm=1,store_md=1 -t $ref -C $sam > $cram";
+        testv $opts, "./test_view -i decode_md=0 -D $cram > $cram.sam_";
+        testv $opts, "./compare_sam.pl  $sam $cram.sam_";
+
+        # Skip auto-MD generation; check MD iff in output file
+        testv $opts, "./test_view -t $ref -C $sam > $cram";
```

Comment:
The bugs in `process_one_read()` are causing all MD:Z to be stored verbatim, whether needed or not.  This test needs to be beefed up to detect that.

Presumably when this PR is working, the `xx#MD.sam` round-tripped with `decode_md=0` should contain no `MD:Z` tags, and `xx#MD2.sam` should have one for each read?